File: data_structures/week3_hash_tables/hash_chain.go
package main

import (
	"fmt"
	"math"
	"strconv"
)

const (
	x = 263
	p = 1000000007
)

type query struct {
	action  string
	content string
}

type hashChain struct {
	size   int
	chains [][]string
}

func RemoveIndex(s []string, index int) []string {
	return append(s[:index], s[index+1:]...)
}

func H(s string, m int) int {
	res, power := 0, 0
	for idx, char := range s {
		power = int(math.Pow(x, float64(idx)))
		res = (res + int(char)*power) % p
	}
	return res % m
}

func (hc hashChain) add(key int, value string) {
	if hc.find(value) {
		return
	}
	hc.chains[key] = append([]string{value}, hc.chains[key]...)
}

func (hc hashChain) check(key int) {
	values := hc.chains[key]
	if len(values) > 0 {
		var s string
		for _, value := range values {
			s += fmt.Sprintf("%s ", value)
		}
		fmt.Println(s)
	} else {
		fmt.Println("")
	}
}

func (hc hashChain) find(content string) bool {
	for _, chain := range hc.chains {
		for _, value := range chain {
			if value == content {
				return true
			}
		}
	}
	return false
}

func (hc hashChain) del(content string) {
	for key, chain := range hc.chains {
		for idx, value := range chain {
			if value == content {
				hc.chains[key] = RemoveIndex(chain, idx)
			}
		}
	}
}

func execQuery(hc hashChain, q query) {
	if q.action == "add" {
		hc.add(H(q.content, hc.size), q.content)
	} else if q.action == "check" {
		key, _ := strconv.Atoi(q.content)
		hc.check(key)
	} else if q.action == "find" {
		if hc.find(q.content) {
			fmt.Println("yes")
		} else {
			fmt.Println("no")
		}
	} else if q.action == "del" {
		hc.del(q.content)
	}
}

func main() {
	var m, num int
	fmt.Scanln(&m)
	fmt.Scanln(&num)

	queries := make([]query, num)
	for i := 0; i < num; i++ {
		var s1, s2 string
		fmt.Scanf("%s %s", &s1, &s2)
		queries[i] = query{action: s1, content: s2}
	}

	hc := hashChain{size: m, chains: make([][]string, m)}
	for _, q := range queries {
		execQuery(hc, q)
	}
	// fmt.Println(hc.chains)
}

File: data_structures/week3_hash_tables/hash_chain.py
# pass all test cases

X = 263
PRIME_NUMBER = 1000000007


class Query:
    def __init__(self, input_string):
        s = input_string.split(' ')
        self.action = s[0]
        self.content = s[1]


class HashChain:
    def __init__(self, m):
        self.size = m
        self.chains = [[] for _ in range(m)]

    def add(self, value):
        key = self.H(value)
        if self.find(value):
            return
        self.chains[key] = [value] + self.chains[key][:]

    def delete(self, value):
        key = self.H(value)
        for i, ele in enumerate(self.chains[key]):
            if ele == value:
                self.chains[key].pop(i)
                return

    def find(self, value):
        if value in self.chains[self.H(value)]:
            return True
        return False

    def check(self, key):
        if self.chains[key]:
            print(' '.join(self.chains[key]))
        else:
            print('')

    def H(self, s):
        res = 0
        for i, char in enumerate(s):
            res = (res + ord(char) * (X**i)) % PRIME_NUMBER
        return res % self.size


def execute_query(hash_chain, query):
    if query.action == 'add':
        hash_chain.add(query.content)
    elif query.action == 'check':
        hash_chain.check(int(query.content))
    elif query.action == 'find':
        if hash_chain.find(query.content):
            print('yes')
        else:
            print('no')
    elif query.action == 'del':
        hash_chain.delete(query.content)


if __name__ == "__main__":
    m = int(input())
    number = int(input())

    queries = []
    for i in range(number):
        queries.append(Query(input()))

    hash_chain = HashChain(m)
    for query in queries:
        execute_query(hash_chain, query)
